src/map/map.ts

```diff
@@ -347,92 +313,114 @@ export class PropertiesMap {
+        }
+
+        const data = this._selected.get(this._active);
+        assert(data !== undefined);
+
+        // Plotly.restyle fires the plotly_click event, so ensure we only run
+        // the update once.
+        // https://github.com/plotly/plotly.js/issues/1025
+        if (data.current === 0 || indexes.environment !== data.current) {
+            data.current = indexes.environment;
+            // Sets the active marker on this map
+            this._updateSelectedMarker(data);
+        }
     }
+
     /**
-     * Function to set the active marker for communicating with the structure
-     * viewer
+     * Set the marker with given GUID as the active marker.
      *
-     * @param  activeGUID the GUID of the new active viewer
+     * @param guid the GUID of the new active viewer
      */
-    public set active(activeGUID: string) {
-        if (this._active === activeGUID) {
-            // guard against infinite recursion
-            // `Structure set active => StructureViewer.show => Map set active
-            // => PropertiesMap.select => Structure set active`
-            return;
+    public setActive(guid: GUID) {
+        if (this._active !== undefined) {
+            const oldData = this._selected.get(this._active);
+            assert(oldData !== undefined);
+            oldData.marker.classList.toggle('chsp-active-structure', false);
         }
 
-        if (!this._selected.has(activeGUID)) {
-          this._addMarker(activeGUID);
+        this._active = guid;
+        const data = this._selected.get(this._active);
+        assert(data !== undefined);
+        data.marker.classList.toggle('chsp-active-structure', true);
+
+        if (this._is3D()) {
+            this._restyle({'marker.size': this._sizes(1)} as Data, 1);
         }
+    }
 
-        if (!this._is3D()) {
-            // if we are in 2D we need to update the visuals
-            if (this._active !== undefined && this._selected.has(this._active)) {
-                // this has been left as document.getElementById for when we are
-                // setting a new active because the old one has been deleted.
-                const oldButton = document.getElementById(`chsp-selected-${this._active}`);
-                if (oldButton !== null) {
-                    oldButton.classList.toggle('chsp-active-structure', false);
-                }
-            }
-            const newButton = getByID(`chsp-selected-${activeGUID}`);
-            this._active = activeGUID;
-            newButton.classList.toggle('chsp-active-structure', true);
+    /**
+     * Add a new marker to the map. The new marker is set as the active one
+     *
+     * @param guid GUID of the new marker
+     * @param indexes indexes of the environment that the new marker should show
+     */
+    public addMarker(guid: GUID, color: string, indexes: Indexes): void {
+        assert(!this._selected.has(guid));
```

Comment:
Why do we need this assertion?

Reply:
To ensure we don't add the same marker multiple times

Reply:
I'm often more in favor of quitting the function than throwing an error, but that's most likely a style choice.

Reply:
The issue with quitting the function is that it can hide programming errors since it fails silently. I try to use [defensive programming](https://en.wikipedia.org/wiki/Defensive_programming) and fail as early as possible to catch control flow errors & other issues